cmd/nodecmd/create.go

```diff
@@ -0,0 +1,678 @@
+	} else {
+		ipInTCP, ipInHTTP := checkCurrentIPInSg(sg, userIPAddress)
+		setSecurityGroupRule(rootBody, userIPAddress, *sg.GroupId, ipInTCP, ipInHTTP)
+	}
+	setElasticIP(rootBody)
+	setUpInstance(rootBody, securityGroupName, useExistingKeyPair, keyPairName, ami)
+	setOutput(rootBody)
+	_, err = tfFile.Write(hclFile.Bytes())
+	if err != nil {
+		return err
+	}
+	instanceID, elasticIP, err := runTerraform()
+	if err != nil {
+		return err
+	}
+	certFilePath, err := getCertFilePath(certName)
+	if err != nil {
+		return err
+	}
+
+	if !useExistingKeyPair {
+		err = handleCerts(certName)
+		if err != nil {
+			return err
+		}
+	}
+
+	inventoryPath := "inventories/" + clusterName
+	if err := createAnsibleHostInventory(inventoryPath, elasticIP, certFilePath); err != nil {
+		return err
+	}
+	time.Sleep(5 * time.Second)
+
+	if err := runAnsiblePlaybook(inventoryPath); err != nil {
+		return err
+	}
+	err = createNodeConfig(instanceID, region, ami, keyPairName, certFilePath, securityGroupName, elasticIP, clusterName)
+	if err != nil {
+		return err
+	}
+	PrintResults(instanceID, elasticIP, certFilePath, region)
+	return nil
+}
+
+func checkKeyPairExists(ec2Svc *ec2.EC2, kpName string) (bool, error) {
+	keyPairInput := &ec2.DescribeKeyPairsInput{
+		KeyNames: []*string{
+			aws.String(kpName),
+		},
+	}
+
+	// Call to get detailed information on each instance
+	_, err := ec2Svc.DescribeKeyPairs(keyPairInput)
+	if err != nil {
+		if strings.Contains(err.Error(), "InvalidKeyPair.NotFound") {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func checkSecurityGroupExists(ec2Svc *ec2.EC2, sgName string) (bool, *ec2.SecurityGroup, error) {
+	sgInput := &ec2.DescribeSecurityGroupsInput{
+		GroupNames: []*string{
+			aws.String(sgName),
+		},
+	}
+
+	sg, err := ec2Svc.DescribeSecurityGroups(sgInput)
+	if err != nil {
+		if strings.Contains(err.Error(), "InvalidGroup.NotFound") {
+			return false, &ec2.SecurityGroup{}, nil
+		}
+		return false, &ec2.SecurityGroup{}, err
+	}
+	return true, sg.SecurityGroups[0], nil
+}
+
+func checkCertInSSHDir(certFilePath string) bool {
+	_, err := os.Stat(certFilePath)
+	return err == nil
+}
+
+func checkCurrentIPInSg(sg *ec2.SecurityGroup, currentIP string) (bool, bool) {
+	var ipInTCP bool
+	var ipInHTTP bool
+	for _, ip := range sg.IpPermissions {
+		if *ip.FromPort == 22 || *ip.FromPort == 9650 {
+			for _, cidrIP := range ip.IpRanges {
+				if strings.Contains(cidrIP.String(), currentIP) {
+					if *ip.FromPort == 22 {
+						ipInTCP = true
+					} else if *ip.FromPort == 9650 {
+						ipInHTTP = true
+					}
+					break
+				}
+			}
+		}
+	}
+	return ipInTCP, ipInHTTP
+}
+
+func setCloudCredentials(rootBody *hclwrite.Body, accessKey, secretKey, region string) error {
+	provider := rootBody.AppendNewBlock("provider", []string{"aws"})
+	providerBody := provider.Body()
+	providerBody.SetAttributeValue("access_key", cty.StringVal(accessKey))
```

Comment:
Does this cause secrets to be written to files unknowningly? Is this the best way to handle auth?

Reply:
We shouldn't be using this pattern to access AWS secrets, there are several methods to do this. We should probably first check environment variables and second do this method https://registry.terraform.io/providers/hashicorp/aws/latest/docs#shared-configuration-and-credentials-files

Reply:
We actually might want to try to match the entire AWS CLI credentials precedence stack https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html#cli-configure-quickstart-precedence

Reply:
As far as how we accept credentials that is. What we don't want to do is expose them in a way that the caller did not intend or expect. I think we really want to avoid writing them to files or anything like that. It'd be easy for a bug to accidentally not delete those files and then the user has credentials lying around.